lib/index.js

```diff
@@ -0,0 +1,148 @@
+    })
+  );
+
+  // GET /payment
+  // GET all payments for an account.
+  app.get(
+    PAYMENT_ENDPOINT,
+    ensureAuthenticated,
+    asyncHandler(async (req, res) => {
+      // this is so a user can not request
+      // another user's payments.
+      const {id: creator} = req.user.account;
+      // check for existing Payments.
+      const query = {creator};
+      const payments = await paymentService.db.findAll({query});
+      return res.status(200).json(payments);
+    })
+  );
+
+  // POST /payment
+  // Creates a new Payment.
+  // The client will need to post the payment they intend to use.
+  app.post(
+    PAYMENT_ENDPOINT,
+    ensureAuthenticated,
+    validate({body: 'payment.creating'}),
+    asyncHandler(async (req, res) => {
+      const {id: creator} = req.user.account;
+      const {payment: paymentData} = req.body;
+      logger.debug('CREATE PAYMENT CALLED', {paymentData});
+      const processingPayments = await paymentService.db.findAll(
+        {query: {creator, status: PaymentStatus.PROCESSING}});
+      // this can cause issues if a payment fails while processing.
+      if(processingPayments.length > 0) {
+        throw new BedrockError(
+          'Can not create a new payment if you have processing payments.',
+          Error.Constraint, {httpStatusCode: 409, public: true});
+      }
+      // check for existing pending Payments.
+      const pendingPayments = await paymentService.db.findAll(
+        {query: {creator, status: PaymentStatus.PENDING}});
+      if(pendingPayments.length > 0) {
+        // this is where most of the concurrency issues appear.
+        const {updatedOrder, payment} = await paymentService.
+          updatePendingPayment({pendingPayments, paymentData});
+        return res.status(200).json({order: updatedOrder, payment});
+      }
+      logger.debug('CREATING NEW PAYMENT', {creator, paymentData});
+      // FIXME remove this try/catch before PR or leave a comment.
+      // this is mostly for debugging and block BedrockErrors from returning.
+      // NOTE: this is being left in for the PR so refactor of bedrock-payments
+      // is easier.
+      try {
+        const {order, payment} = await paymentService.createPayment(
+          {creator, paymentData});
+        return res.status(201).json({order, payment});
+      } catch(e) {
+        logger.error('FAILED TO CREATE PAYPAL ORDER ERROR', {error: e});
+        res.sendStatus(500);
+      }
+    }));
+
+  // PUT /payment/:paymentId.
+  // Processes an existing payment.
+  app.put(
+    `${PAYMENT_ENDPOINT}/:paymentId`,
+    ensureAuthenticated,
+    validate({body: 'payment.processing'}),
+    asyncHandler(async (req, res) => {
+      logger.debug('PROCESS PAYMENT CALLED', {body: req.body});
+      const {id: account} = req.user.account;
+      const creator = account;
+      const {payment: paymentData, order = null} = req.body;
+      // get the payment
+      const payment = await paymentService.db.findOne(
```

Comment:
can the `creator` be added to this query to prevent ever retrieving a payment for the wrong user, and therefore we don't have to check for that error condition later.

Reply:
This is a good idea, and I implemented it, but it does result in a more vague 404. That might be better as the account trying to use another account's payment gets a 404 instead of a 401. Less information for the attacker.